connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java

```diff
@@ -431,15 +427,29 @@ private <T> Collection<PluginDesc<T>> getServiceLoaderPluginDesc(Class<T> klass,
                     log.error("Failed to discover {}{}", klass.getSimpleName(), reflectiveErrorDescription(t.getCause()), t);
                     continue;
                 }
-                result.add(pluginDesc((Class<? extends T>) pluginImpl.getClass(),
-                    versionFor(pluginImpl), loader));
+                Class<? extends T> pluginKlass = (Class<? extends T>) pluginImpl.getClass();
+                if (!isParentClassloader(pluginKlass.getClassLoader(), loader)) {
+                    log.debug("Exclude {} that is from classloader {}", pluginKlass.getSimpleName(), pluginKlass.getClassLoader());
+                    continue;
+                }
+                result.add(pluginDesc(pluginKlass, versionFor(pluginImpl), loader));
             }
         } finally {
             Plugins.compareAndSwapLoaders(savedLoader);
         }
         return result;
     }
 
+    private static boolean isParentClassloader(ClassLoader loader, ClassLoader parent) {
+        while (loader != null) {
+            if (loader == parent) {
+                return true;
+            }
+            loader = loader.getParent();
+        }
+        return false;
```

Comment:
I wrote this in case a plugin came from a classloader which was a child of the PluginClassLoader. Thinking about it again, I don't know how a plugin would manage that, since the PluginClassLoader doesn't have any child-first logic that would permit it to load classes from deeper in the classloading hierarchy.

I'll simplify this logic.